Stashed bad command during replay causes JournalFailureException

Noted in a test where I expected to be able to send a message that causes a class cast exception, but on travis which is slower than my workstation, it is still replaying when that message is sent, and the message is stashed, to be unstashed when replay completes, however when it then fails it is reported as a JournalFailureException:

```
akka.persistence.typed.internal.JournalFailureException: Exception during recovery.  Last known sequence number [0]. PersistenceId [foiesftr]
	at akka.persistence.typed.internal.ReplayingEvents.onRecoveryFailure(ReplayingEvents.scala:167)
	at akka.persistence.typed.internal.ReplayingEvents.onJournalResponse(ReplayingEvents.scala:108)
	at akka.persistence.typed.internal.ReplayingEvents.$anonfun$stay$1(ReplayingEvents.scala:70)
	at akka.actor.typed.internal.BehaviorImpl$ReceiveMessageBehavior.receive(BehaviorImpl.scala:49)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:421)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:394)
	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:49)
	at akka.actor.typed.internal.WidenedInterceptor.aroundReceive(InterceptorImpl.scala:168)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:73)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:421)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:394)
	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:49)
	at akka.persistence.typed.internal.EventSourcedBehaviorImpl$$anon$1.aroundReceive(EventSourcedBehaviorImpl.scala:107)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:73)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:421)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:394)
	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:49)
	at akka.actor.typed.internal.SimpleSupervisor.aroundReceive(Supervision.scala:85)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:73)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:421)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:394)
	at akka.actor.typed.internal.adapter.ActorAdapter.akka$actor$typed$internal$adapter$ActorAdapter$$handleMessage(ActorAdapter.scala:82)
	at akka.actor.typed.internal.adapter.ActorAdapter$$anonfun$running$1.applyOrElse(ActorAdapter.scala:78)
	at akka.actor.Actor.aroundReceive(Actor.scala:517)
	at akka.actor.Actor.aroundReceive$(Actor.scala:515)
	at akka.actor.typed.internal.adapter.ActorAdapter.aroundReceive(ActorAdapter.scala:39)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:592)
	at akka.actor.ActorCell.invoke(ActorCell.scala:561)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)
	at akka.dispatch.Mailbox.run(Mailbox.scala:225)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.String
	at akka.persistence.typed.javadsl.Effect.$anonfun$thenRun$1(Effect.scala:102)
	at akka.persistence.typed.javadsl.Effect.$anonfun$thenRun$1$adapted(Effect.scala:102)
	at akka.persistence.typed.internal.Running.applySideEffect(Running.scala:319)
	at akka.persistence.typed.internal.Running.applySideEffects(Running.scala:304)
	at akka.persistence.typed.internal.Running.applyEffects$1(Running.scala:141)
	at akka.persistence.typed.internal.Running.onCommand$1(Running.scala:83)
	at akka.persistence.typed.internal.Running.$anonfun$handlingCommands$3(Running.scala:161)
	at akka.actor.typed.internal.BehaviorImpl$ReceiveMessageBehavior.receive(BehaviorImpl.scala:49)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:421)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:394)
	at akka.actor.typed.Behavior$.interpretOne$1(Behavior.scala:440)
	at akka.actor.typed.Behavior$.interpretMessages(Behavior.scala:446)
	at akka.actor.typed.internal.StashBufferImpl.unstash(StashBufferImpl.scala:103)
	at akka.persistence.typed.internal.StashManagement.tryUnstash(StashManagement.scala:53)
	at akka.persistence.typed.internal.StashManagement.tryUnstash$(StashManagement.scala:48)
	at akka.persistence.typed.internal.ReplayingEvents.tryUnstash(ReplayingEvents.scala:52)
	at akka.persistence.typed.internal.ReplayingEvents.onRecoveryCompleted(ReplayingEvents.scala:182)
	at akka.persistence.typed.internal.ReplayingEvents.onJournalResponse(ReplayingEvents.scala:98)
	... 33 more
```

--- a/akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala
+++ b/akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala
@@ -5,6 +5,7 @@
 package akka.persistence.typed.internal
 
 import scala.util.control.NonFatal
+import scala.util.control.NoStackTrace
 
 import akka.actor.typed.Behavior
 import akka.actor.typed.internal.PoisonPill
@@ -13,6 +14,7 @@ import akka.annotation.InternalApi
 import akka.event.Logging
 import akka.persistence.JournalProtocol._
 import akka.persistence._
+import akka.persistence.typed.internal.ReplayingEvents.FailureWhileUnstashing
 
 /***
  * INTERNAL API
@@ -45,6 +47,8 @@ private[akka] object ReplayingEvents {
     state: ReplayingState[S]
   ): Behavior[InternalProtocol] =
     new ReplayingEvents(setup.setMdc(MDC.ReplayingEvents)).createBehavior(state)
+
+  private final case class FailureWhileUnstashing(cause: Throwable) extends Exception(cause) with NoStackTrace
 
 }
 
@@ -104,6 +108,7 @@ private[akka] class ReplayingEvents[C, E, S](override val setup: BehaviorSetup[C
           Behaviors.unhandled
       }
     } catch {
+      case FailureWhileUnstashing(ex) ⇒ throw ex
       case NonFatal(cause) ⇒
         onRecoveryFailure(cause, state.seqNr, None)
     }
@@ -179,7 +184,11 @@ private[akka] class ReplayingEvents[C, E, S](override val setup: BehaviorSetup[C
         Running.RunningState[S](state.seqNr, state.state, state.receivedPoisonPill)
       )
 
-      tryUnstashOne(running)
+      try {
+        tryUnstashOne(running)
+      } catch {
+        case NonFatal(t) ⇒ throw FailureWhileUnstashing(t)
+      }
     }
   } finally {
     setup.cancelRecoveryTimer()

--- a/akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedBehaviorSpec.scala
+++ b/akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedBehaviorSpec.scala
@@ -12,9 +12,8 @@ import scala.concurrent.Promise
 import scala.concurrent.duration._
 import scala.util.Success
 import scala.util.Try
-
 import akka.Done
-import akka.actor.testkit.typed.TestKitSettings
+import akka.actor.testkit.typed.{TestException, TestKitSettings}
 import akka.actor.testkit.typed.scaladsl._
 import akka.actor.typed.ActorRef
 import akka.actor.typed.ActorSystem
@@ -72,6 +71,7 @@ object EventSourcedBehaviorSpec {
   def conf: Config = ConfigFactory.parseString(
     s"""
     akka.loglevel = INFO
+    akka.loggers = [akka.testkit.TestEventListener]
     # akka.persistence.typed.log-stashing = on
     akka.persistence.journal.leveldb.dir = "target/typed-persistence-${UUID.randomUUID().toString}"
     akka.persistence.journal.plugin = "akka.persistence.journal.leveldb"
@@ -100,6 +100,8 @@ object EventSourcedBehaviorSpec {
   final case object DelayFinished extends Command
   private case object Timeout extends Command
   final case object LogThenStop extends Command
+  final case object Fail extends Command
+  final case object StopIt extends Command
 
   sealed trait Event
   final case class Incremented(delta: Int) extends Event
@@ -164,7 +166,7 @@ object EventSourcedBehaviorSpec {
 
         case cmd: IncrementWithConfirmation ⇒
           Effect.persist(Incremented(1))
-            .thenReply(cmd)(newState ⇒ Done)
+            .thenReply(cmd)(_ ⇒ Done)
 
         case GetValue(replyTo) ⇒
           replyTo ! state
@@ -222,6 +224,13 @@ object EventSourcedBehaviorSpec {
               loggingActor ! firstLogging
             }
             .thenStop
+
+        case Fail ⇒
+          throw new TestException("boom!")
+
+        case StopIt ⇒
+          Effect.none.thenStop()
+
       },
       eventHandler = (state, evt) ⇒ evt match {
         case Incremented(delta) ⇒
@@ -247,6 +256,9 @@ class EventSourcedBehaviorSpec extends ScalaTestWithActorTestKit(EventSourcedBeh
   implicit val materializer = ActorMaterializer()(system.toUntyped)
   val queries: LeveldbReadJournal = PersistenceQuery(system.toUntyped).readJournalFor[LeveldbReadJournal](
     LeveldbReadJournal.Identifier)
+
+  // needed for the untyped event filter
+  implicit val actorSystem = system.toUntyped
 
   val pidCounter = new AtomicInteger(0)
   private def nextPid(): PersistenceId = PersistenceId(s"c${pidCounter.incrementAndGet()})")
@@ -650,15 +662,38 @@ class EventSourcedBehaviorSpec extends ScalaTestWithActorTestKit(EventSourcedBeh
     }
 
     "fail after recovery timeout" in {
-      val c = spawn(Behaviors.setup[Command](ctx ⇒
-        counter(ctx, nextPid)
-          .withSnapshotPluginId("slow-snapshot-store")
-          .withJournalPluginId("short-recovery-timeout"))
-      )
+      EventFilter.error(start = "Persistence failure when replaying snapshot", occurrences = 1).intercept {
+        val c = spawn(Behaviors.setup[Command](ctx ⇒
+          counter(ctx, nextPid)
+            .withSnapshotPluginId("slow-snapshot-store")
+            .withJournalPluginId("short-recovery-timeout"))
+        )
 
-      val probe = TestProbe[State]
+        val probe = TestProbe[State]
 
-      probe.expectTerminated(c, probe.remainingOrDefault)
+        probe.expectTerminated(c, probe.remainingOrDefault)
+      }
+    }
+
+    "not wrap a failure caused by command stashed while recovering in a journal failure" in {
+      val pid = nextPid()
+      val probe = TestProbe[AnyRef]
+
+      // put some events in there, so that recovering takes a little time
+      val c = spawn(Behaviors.setup[Command](counter(_, pid)))
+      (0 to 50).foreach { _ ⇒
+        c ! IncrementWithConfirmation(probe.ref)
+        probe.expectMessage(Done)
+      }
+      c ! StopIt
+      probe.expectTerminated(c)
+
+      EventFilter[TestException](occurrences = 1).intercept {
+        val c2 = spawn(Behaviors.setup[Command](counter(_, pid)))
+        c2 ! Fail
+        probe.expectTerminated(c2) // should fail
+      }
+
     }
 
     def watcher(toWatch: ActorRef[_]): TestProbe[String] = {
@@ -667,7 +702,7 @@ class EventSourcedBehaviorSpec extends ScalaTestWithActorTestKit(EventSourcedBeh
         ctx.watch(toWatch)
         Behaviors.receive[Any] { (_, _) ⇒ Behaviors.same }
           .receiveSignal {
-            case (_, s: Terminated) ⇒
+            case (_, _: Terminated) ⇒
               probe.ref ! "Terminated"
               Behaviors.stopped
           }

--- a/akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedBehaviorSpec.scala
+++ b/akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedBehaviorSpec.scala
@@ -13,7 +13,8 @@ import scala.concurrent.duration._
 import scala.util.Success
 import scala.util.Try
 import akka.Done
-import akka.actor.testkit.typed.{TestException, TestKitSettings}
+import akka.testkit.EventFilter
+import akka.actor.testkit.typed.{ TestException, TestKitSettings }
 import akka.actor.testkit.typed.scaladsl._
 import akka.actor.typed.ActorRef
 import akka.actor.typed.ActorSystem

